67_RayQueryGeometry/main.cpp

```diff
@@ -0,0 +1,1010 @@
+			objectsCpu[OT_ICOSPHERE] = ReferenceObjectCpu{ .meta = {.type = OT_ICOSPHERE, .name = "Icosphere Mesh" }, .shadersType = GP_ICO, .data = gc->createIcoSphere(1, 3, true) };
+
+			for (uint32_t i = 0; i < objectsCpu.size(); i++)
+			{
+				const auto& geom = objectsCpu[i];
+				auto& obj = objectsGpu[i];
+
+				obj.meta.name = geom.meta.name;
+				obj.meta.type = geom.meta.type;
+
+				obj.indexCount = geom.data.indexCount;
+				obj.indexType = geom.data.indexType;
+
+				auto vBuffer = smart_refctd_ptr(geom.data.bindings[0].buffer); // no offset
+				IGPUBuffer::SCreationParams vParams;
+				vParams.size = vBuffer->getSize();
+				vParams.usage = bitflag(asset::IBuffer::EUF_STORAGE_BUFFER_BIT) | asset::IBuffer::EUF_TRANSFER_DST_BIT | asset::IBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF | 
+					asset::IBuffer::EUF_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT | asset::IBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;
+				obj.bindings.vertex.offset = 0u;
+				auto vertexBuffer = m_device->createBuffer(std::move(vParams));
+
+				//if (!vertexBuffer)
+				//	return false;
+
+				auto iBuffer = smart_refctd_ptr(geom.data.indexBuffer.buffer); // no offset
+				IGPUBuffer::SCreationParams iParams;
+				iParams.size = iBuffer->getSize();
+				iParams.usage = bitflag(asset::IBuffer::EUF_STORAGE_BUFFER_BIT) | asset::IBuffer::EUF_TRANSFER_DST_BIT | asset::IBuffer::EUF_INLINE_UPDATE_VIA_CMDBUF |
+					asset::IBuffer::EUF_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT | asset::IBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;
+				obj.bindings.index.offset = 0u;
+				auto indexBuffer = m_device->createBuffer(std::move(iParams));
+
+				//if (geom.data.indexType != EIT_UNKNOWN)
+				//	if (!indexBuffer)
+				//		return false;
+
+				for (auto buf : { vertexBuffer, indexBuffer })
+				{
+					if (buf)
+					{
+						auto reqs = buf->getMemoryReqs();
+						reqs.memoryTypeBits &= m_device->getPhysicalDevice()->getUpStreamingMemoryTypeBits();
+						m_device->allocate(reqs, buf.get(), IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT);
+					}
+				}
+
+				// TODO: couldn't figure out how to use SIntendedNextSubmit and updateBufferRangeViaStagingBuffer
+				obj.bindings.vertex = { .offset = 0u, .buffer = std::move(vertexBuffer) };
+				SBufferRange<IGPUBuffer> vRange = { .offset = obj.bindings.vertex.offset, .size = obj.bindings.vertex.buffer->getSize(), .buffer = obj.bindings.vertex.buffer };
+				cmdbuf->updateBuffer(vRange, vBuffer->getPointer());
+
+				obj.bindings.index = { .offset = 0u, .buffer = std::move(indexBuffer) };
+				SBufferRange<IGPUBuffer> iRange = { .offset = obj.bindings.index.offset, .size = obj.bindings.index.buffer->getSize(), .buffer = obj.bindings.index.buffer };
+				cmdbuf->updateBuffer(iRange, iBuffer->getPointer());
+			}
+
+			return true;
+		}
+
+		bool createAccelerationStructures(IGPUCommandBuffer* cmdbuf)
+		{
+			// build bottom level ASes
+			{
+				const auto& obj = objectsGpu[OT_CUBE];
```

Comment:
when you progress to do all objects, do 3 waves (so you have most overlap/least pipeline barriers):
1. build all BLASes
2. compact all BLASes
3. build the TLAS